### PR TITLE
feat(integration): Add Anthropic Messages.stream support

### PR DESCRIPTION
**Pull request by tcapelle**
Antrhopic has 2 types of streaming: One by event and another by messages.

We are missing the `Messages.stream` support, this PR tries to add this. On this mode, the antrhopic library gives you back a ContextManager:
```python
class MessageStreamManager:
    """Wrapper over MessageStream that is returned by `.stream()`.

    ```py
    with client.messages.stream(...) as stream:
        for chunk in stream:
            ...
    ```
    """

    def __init__(
        self,
        api_request: Callable[[], Stream[RawMessageStreamEvent]],
    ) -> None:
        self.__stream: MessageStream | None = None
        self.__api_request = api_request

    def __enter__(self) -> MessageStream:
        raw_stream = self.__api_request()

        self.__stream = MessageStream(
            cast_to=raw_stream._cast_to,
            response=raw_stream.response,
            client=raw_stream._client,
        )

        return self.__stream

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None:
        if self.__stream is not None:
            self.__stream.close()
```

I am curious if patching the `_IteratorWrapper` with this hack:
```python
    def __enter__(self) -> "_IteratorWrapper":
        if hasattr(self._iterator, "__enter__"):
            return self._iterator.__enter__()
        return self
```
it is a good idea.

### CODE REVIEW

**Comment by tcapelle** [comment]
cc @tssweeney

File: weave/integrations/anthropic/anthropic_sdk.py [hunk 1]
```diff
@@ -103,12 +102,38 @@ async def _async_wrapper(
...
     return wrapper
 
 
+def anthropic_stream_accumulator(
+    acc: typing.Optional["Message"],
+    value: "MessageStream",
+) -> "Message":
+    from anthropic.lib.streaming._types import MessageStopEvent
+    
+    print(f"    type(value): {type(value)}")
```
Comment (tssweeney) [inline weave/integrations/anthropic/anthropic_sdk.py:116]: please remove print

File: weave/integrations/anthropic/anthropic_sdk.py [hunk 2]
```diff
@@ -103,12 +102,38 @@ async def _async_wrapper(
...
+    if acc is None:
+        acc = {}
+    if isinstance(value, MessageStopEvent):
+        print(f"Here we go!")
```
Comment (tssweeney) [inline weave/integrations/anthropic/anthropic_sdk.py:120]: also here

File: weave/trace/op_extensions/accumulator.py [hunk 3]
```diff
@@ -219,6 +221,8 @@ def __exit__(
         exc_value: Optional[BaseException],
         traceback: Optional[Any],
     ) -> None:
+        if hasattr(self._iterator, "__exit__"):
+            self._iterator.__exit__(exc_type, exc_value, traceback)
```
Comment (tssweeney) [inline weave/trace/op_extensions/accumulator.py:225]: you might want to put this at the end of the method

File: weave/integrations/anthropic/anthropic_sdk.py [hunk 4]
```diff
@@ -103,12 +102,38 @@ async def _async_wrapper(
...
+        acc = value
```
Comment (tssweeney) [inline weave/integrations/anthropic/anthropic_sdk.py:121]: does this imply the entire payload is here?

Comment (tcapelle) [reply weave/integrations/anthropic/anthropic_sdk.py:121]: yeap =)

File: weave/integrations/anthropic/anthropic_sdk.py (hunk 1 above)
Comment (tcapelle) [reply weave/integrations/anthropic/anthropic_sdk.py:116]: will remove once I we are are ready

**Comment by tcapelle** [comment]
Ok, now it is working but...

- Had to change some things in the `accumulator.py` like being able to subclass the `_IteratorWrapper`
- This is kind of brittle, but I suppose all the integrations are brittle.
- Couldn't find a way to deal with all the possible streaming scenarios that this lib gives

**Comment by tcapelle** [comment]
I think this is ready for review @tssweeney

File: weave/trace/op_extensions/accumulator.py [hunk 5]
```diff
@@ -130,12 +31,12 @@ class _IteratorWrapper(Generic[V]):
 
     def __init__(
         self,
-        iterator: Union[Iterator, AsyncIterator],
+        iterator_or_ctx_manager: Union[Iterator, AsyncIterator],
```
Comment (andrewtruong) [inline weave/trace/op_extensions/accumulator.py:34]: Can you help me understand why you need this?

**Review by andrewtruong — APPROVED** [review]

**Pull request merged**